Draft R dev blog post (#400)
RConsortium/S7: vignettes/R-dev-blog-post.qmd

This blog post outlines four patches that have been integrated into base R to facilitate the existence of S7 as a standalone R package. Since S7 is built on top of S3, the new capabilities are available not only to S7 but also to all packages built on top of S3.

Each of these four patches introduces a new S3 generic:

-   `chooseOpsMethod()`
-   `nameOfClass()`
-   `@`
-   `%*%`

This blog post describes these new generics, the features in S7 that motivated their implementation, and their utility to the broader R community outside of S7. These enhancements are generally beneficial for packages integrating or implementing alternative OOP systems. To illustrate these additional benefits, we use the reticulate package as an example. Reticulate bridges Python's OOP semantics with R's S3 system.

### `chooseOpsMethod()`

> S3 only supports dispatching generics on their first argument, while S4 supports dispatching generics on an arbitrary number of arguments. S7 uses S3 dispatch, but also, implements multiple dispatch. To enable both S3 support and multiple dispatch, dispatch in S7 is performed in two stages: S3 generics first dispatch on the first method to an S7 aware method, which then invokes S7-internal dispatch routines that support multiple dispatch. These are implementation details which, when everything works, are hidden from S7 users. The complication which motivated the addition of a new S3 generic is the Ops Group.

Reviewer: ```suggestion
S3 only supports dispatching generics on their first argument, while S4 supports dispatching generics on an arbitrary number of arguments. S7 uses S3 dispatch, but also implements its own double dispatch (i.e. multiple dispatch on exactly two arguments). To enable both S3 support and multiple dispatch, dispatch in S7 is performed in two stages: S3 generics first dispatch on the first method to an S7 aware method, which then invokes S7-internal dispatch routines that support multiple dispatch. These are implementation details which, when everything works, are hidden from S7 users. The complication which motivated the addition of a new S3 generic is the Ops Group.
```

Unlike other S3 generics, the Ops Group Generics, consisting of infix binary operators like `+`, `-`, `*`, and so on, support dispatching to S3 methods on either the first or second argument. Before R 4.3.0, if both objects in the first and second arguments implemented a method for the Ops Group Generic, R would signal an error. Starting with R 4.3.0, when both the first and second argument to an Ops generic implement a suitable method, R will invoke `chooseOpsMethod()` instead of signaling an error, giving an opportunity for the object types to declare that their Ops method implementation can handle the combination. If the `chooseOpsMethod()` method for an object returns `TRUE`, then the method found for that object is used by the Ops generic.

Because S7 implements its own generic dispatch mechanism, the `chooseOpsMethod()` method for S7 object always returns `TRUE`. This permits, for example, an S7 object type to define a `+` method that works with other S3 objects, like `Sys.Date()`.

```{r, eval = FALSE}
library(S7)

ClassX <- new_class("ClassX")

method(`+`, list(ClassX, class_any)) <-
  function(e1, e2) "method: X + class_any"

x <- ClassX()

x + Sys.Date()
```

> This also has utility for other packages that implement alternative OO systems. For example, reticulate is an R package that embeds Python in an R session, and allows R users to operate with Python objects directly. Python has its own implementation of method dispatch for infix operators like `+`.
> 
> The addition of `chooseOpsMethod()` now allows both arguments in a call like `a + b` to implement a suitable `+` method. Specifically, it enables reticulate to implement a default fallback `+` method that dispatches to the corresponding Python routines, while also allowing R users to define a custom `+` method for specific Python object types. For example, the `+` operator can now select the appropriate method if methods for TensorFlow tensors and Python objects generally are both be implemented. Previously, the restriction of only one of the arguments implementing an S3 method required that the second argument in an Ops call not be an S3 class.

Reviewer: ```suggestion
This also has utility for other packages that implement alternative OO systems. For example, reticulate is an R package that embeds Python in an R session, and allows R users to operate with Python objects directly. Python has its own implementation of method dispatch for infix operators like `+`. The addition of `chooseOpsMethod()` enables reticulate to implement a default fallback `+` method that dispatches to the corresponding Python routines, while still allowing R users to define a custom `+` method for specific Python object types. 

For example, the `+` operator can now select the appropriate method if methods for TensorFlow tensors and Python objects generally are both be implemented. Previously, the restriction of only one of the arguments implementing an S3 method required that the second argument in an Ops call not be an S3 class.
```

```{r, eval = FALSE}
library(reticulate)

# Adding a NumPy array and an R array dispatches to the default
# `+` method for "python.builtin.object"
np_array(1:3) + array(1:3)

# Adding a TensorFlow Tensor and an R array invokes
# invokes the specific `+` method for "tensorflow.tensor"
array(1:3) + tensorflow::as_tensor(1:3)

# Adding a NumPy array with a TensorFlow Tensor.
# Prior to R 4.3.0, this would signal an error, Warning about
#   Incompatible methods ("+.tensorflow.tensor", "+.python.builtin.object")
# Beginning with R 4.3.0, chooseOpsMethod() is called to choose the appropriate
# Ops method, and this now works
np_array(1:3) + as_tensor(1:3)
```

### `nameOfClass()`

Though S7 uses S3 dispatch, the class name used for S3 dispatch is considered an implementation detail (it is non-syntatic and mangled to minimize the potential for conflicts, and not intended to be user-facing). Additionally, it can be subject to change if the generic definition migrates between packages.

> For these reasons, it is desirable to be able to check for inheritance without directly using the S3 class name. To enable this, beginning with R 4.3.0 `base::inherits()` now accepts arbitrary objects for the second argument. If the second argument is an S3 object, the new `nameOfClass()` S3 generic is invoked to resolve the appropriate class name.

Reviewer: ```suggestion
For these reasons, it is desirable to be able to check for inheritance without directly using the S3 class name. To enable this, beginning with R 4.3.0 `base::inherits()` now accepts arbitrary objects for the second argument: if the second argument is an S3 object, the new `nameOfClass()` S3 generic is invoked to resolve the appropriate class name.
```

This permits usage in S7 like this:

```{r, eval = FALSE}
ClassX <- S7::new_class("ClassX")

x <- ClassX()

inherits(x, ClassX)
```

This also has utility for other R packages, in particular, packages where an external class hierarchy is being mirrored in R. For example, in reticulate, the S3 class vector for Python object is generated from the Python module where the class is defined. However, in sophisticated codebases like Keras and TensorFlow, the location where a class is defined in the Python sources is considered an internal implementation detail, and often changes between library versions (i.e., the module where a class is defined is decoupled from the module where users are meant to access the class). To ensure compatibility between and across Python library versions, reticulate now implements a `nameOfClass()` method which enables R usage like this:

```{r, eval = FALSE}
inherits(x, tensorflow::tf$Variable)
```

### `@`

> Prior to R 4.3.0, `@` could be used only with S4 objects. Beginning with R 4.3.0, this operator can now perform dispatch S3 dispatch as well. This bring `@` to parity with `@<-` (which has long been able to do S3 dispatch), and makes `@` a peer (in dispatch-ability) of `$`. This is a strictly backwards compatible change, as S3 dispatch is only performed *after* S4 routines have had an opportunity to dispatch.

Reviewer: Maybe note that there's still no default method for `@` but it's available for use in packages?

> The ability to have formally defined properties is one of the benefits of S4; it enables collaborative work at a greater scale than the ad-hoc conventions of S3 classes, and having a dedicated operator for property access in S7 is desirable. Other than performing S3 dispatch, there are no other user-facing changes with `@` in base R.

Reviewer: ```suggestion
The ability to have formally defined properties is one of the benefits of S4 and S7; it enables collaborative work at a greater scale than the ad-hoc conventions of S3 classes, and having a dedicated operator for property access in S7 is desirable. Other than performing S3 dispatch, there are no other user-facing changes with `@` in base R.
```
